src/OAuth/OAuth2/Service/BattleNetCN.php

```diff
@@ -0,0 +1,18 @@
+<?php
+namespace OAuth\OAuth2\Service;
+
+use OAuth\Common\Http\Uri\Uri;
+use OAuth\Common\Consumer\Credentials;
+use OAuth\Common\Http\Client\ClientInterface;
+use OAuth\Common\Storage\TokenStorageInterface;
+use OAuth\Common\Http\Uri\UriInterface;
+
+class BattleNetCN extends BattleNetBase
+{
+    public function __construct(Credentials $credentials, ClientInterface $httpClient, TokenStorageInterface $storage, $scopes = array(), UriInterface $baseApiUri = null)
```

Comment:
Ok. That's the part I was missing. My php is marginal as I'm primarily a c# developer. As long as it can be passed, then that makes more sense (as I mentioned in my original pull request).  I will make the update when I have the time unless someone else gets a chance before then. My time for this is pretty limited right now unfortunately. 


Sent from my Verizon Wireless 4G LTE smartphone

-------- Original message --------
From: Elliot Chance <notifications@github.com> 
Date: 09/09/2015  8:21 PM  (GMT-05:00) 
To: Lusitanian/PHPoAuthLib <PHPoAuthLib@noreply.github.com> 
Cc: AJH16 <ajh16@nycap.rr.com> 
Subject: Re: [PHPoAuthLib] Added Battle.Net Service (#387) 

In src/OAuth/OAuth2/Service/BattleNetCN.php:
> @@ -0,0 +1,18 @@
> +<?php
> +namespace OAuth\OAuth2\Service;
> +
> +use OAuth\Common\Http\Uri\Uri;
> +use OAuth\Common\Consumer\Credentials;
> +use OAuth\Common\Http\Client\ClientInterface;
> +use OAuth\Common\Storage\TokenStorageInterface;
> +use OAuth\Common\Http\Uri\UriInterface;
> +
> +class BattleNetCN extends BattleNetBase
> +{
> +    public function __construct(Credentials $credentials, ClientInterface $httpClient, TokenStorageInterface $storage, $scopes = array(), UriInterface $baseApiUri = null)

I understand what your trying to achieve, using the service factory to generate a service that is already pointed to the correct region. We have this ability already through the ServiceFactory:

$battleNet = $serviceFactory->createService(
    'BattleNet',
    $credentials,
    $storage,
    []
    new Uri(BattleNet::REGION_CN)
);

Then you may enforce requirements that this must be provided quite easily by validating that $baseApiUri is not null in your constructor.

Or if the developer wanted this level of abstraction they could create their own BattleNetServiceFactory.

The region is just a configuration parameter of the service, much like the environment of some other services carry their own URL and must be set in the same way. Its not appropriate to create different services that are all really the same service with slightly different configuration.

—
Reply to this email directly or view it on GitHub.